### PR TITLE
test(engine): fix executor manager heartbeat timing bug in unit test

### LINKED ISSUE #5739: Unit test TestExecutorManagerWatch is not stable
### Which jobs are flaking?

verify CI

### Which test(s) are flaking?

dataflow engine unit test

FAIL: engine/servermaster TestExecutorManagerWatch

### Jenkins logs or GitHub Actions link

https://ci.pingcap.net/blue/organizations/jenkins/ut-check/detail/ut-check/15299/pipeline/

```bash
[2022-06-03T08:55:27.530Z] === Failed

[2022-06-03T08:55:27.530Z] === FAIL: engine/servermaster TestExecutorManagerWatch (2.01s)

[2022-06-03T08:55:27.530Z]     executor_manager_test.go:139: 

[2022-06-03T08:55:27.530Z]         	Error Trace:	executor_manager_test.go:139

[2022-06-03T08:55:27.530Z]         	            				asm_amd64.s:1571

[2022-06-03T08:55:27.530Z]         	Error:      	Expected nil, but got: &enginepb.Error{Code:5, Message:"[DFLOW:ErrTombstoneExecutor]executor 6b1fd956-2063-4e46-9123-c6e1772cce16 has been dead", NotLeader:(*enginepb.NotLeader)(nil)}

[2022-06-03T08:55:27.530Z]         	Test:       	TestExecutorManagerWatch


[2022-06-03T08:55:27.531Z]     executor_manager_test.go:136: 

[2022-06-03T08:55:27.531Z]         	Error Trace:	executor_manager_test.go:136

[2022-06-03T08:55:27.531Z]         	Error:      	Condition never satisfied

[2022-06-03T08:55:27.531Z]         	Test:       	TestExecutorManagerWatch
```

### Anything else we need to know

- Does this test exist for other branches as well?

- Has there been a high frequency of failure lately?

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve?

Issue Number: close #5739

### What is changed and how it works?

The unstable test fails because executor-2 could be offline before it sends a new heartbeat.

- Add a background heartbeat goroutine to send heartbeat periodically 
- Fix a data race in executor manager.

#### Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/servermaster/executor_manager.go [hunk 1]
```diff
@@ -285,9 +285,11 @@ func (e *ExecutorManagerImpl) ExecutorCount(status model.ExecutorStatus) (count
 	e.mu.Lock()
 	defer e.mu.Unlock()
 	for _, executor := range e.executors {
+		executor.mu.RLock()
```
Comment (maxshuang) [inline engine/servermaster/executor_manager.go:288]: It seems that `Executor` is in the boundary of `ExecutorManagerImpl`, which means the managerment and lock protection should be the responsible of `ExecutorManagerImpl`.

Any case show that the state of `Executor` will be updated independently?

Comment (amyangfei) [reply engine/servermaster/executor_manager.go:288]: https://github.com/pingcap/tiflow/blob/d8afd07f45f5f38be1d89ccb3447b4aa70b44177/engine/servermaster/executor_manager.go#L131-L140

**Review by maxshuang — APPROVED** [review]

File: engine/servermaster/executor_manager.go (hunk 1 above)
Comment (maxshuang) [reply engine/servermaster/executor_manager.go:288]: Can you wrap these logic to the method of `Executor`?
It's easy for developer not hold the `exec.mu` when update the status of `Executor`.

**Pull request merged**